src/LasagnaController.cpp

```diff
@@ -1,29 +1,52 @@
 #include "LasagnaController.h"
 
 LasagnaController::LasagnaController() :
-  v1_filter({{1.0003141592601912L, -1.9999999013039569L, 0.99968584073980871L}, {4.93480216e-07L,   4.93480216e-07L,  -4.93480216e-07L,  -4.93480216e-07L}}),
-  v2_filter({{1.0007479981811036L, -1.9999994404986428L, 0.9992520018188963L}, {2.79750679e-06L,  2.79750679e-06L, -2.79750679e-06L, -2.79750679e-06L}}),
-  action_filter({{1L, -0.99994L, 0.0L}, {0.5L, 0.5L, -0.0L}})
+  v1_filter(1./60./16., 0.5, 20.),
+  v2_filter(1./60./7.,  0.5, 20.),
+  action_filter(1./60./16., 0.5, 20.)
 {}
 
+LasagnaController::LasagnaController(float freq) : 
+  v1_filter(1./60./16., 0.5, freq),
+  v2_filter(1./60./7.,  0.5, freq),
+  action_filter(1./60./16., 0.5, freq) 
+{
+  constants.freq = freq;
+}
+
 bool LasagnaController::update(Input input){ 
 
+  /**
+   * Input protection for dtdl_ext
+   */
   if(isnan(input.dldt_ext)){
     input.dldt_ext = 0; 
   } else {
-    float limit = 0.0001;
+    float limit = 0.0001*4;
     input.dldt_ext = pasta_clamp(input.dldt_ext,-limit,limit);
   }
 
+  /**
+   * Update both filtered velocities
+   */
+  if(!is_first_call){ 
+    state.v1 = v1_filter.update((input.h_rel-state.h_rel_last)*constants.freq);
+    state.v2 = v2_filter.update((input.h_rel-state.h_rel_last)*constants.freq);
+  }
+
+  /**
+   * Cases and transistions for various flight states
+   */
   switch(state.status){
     case PRELAUNCH:
-      if(launch_h > constants.equil_h_thresh){        // this is the first time lasagna is called on startup
+      if(is_first_call){        // this is the first time lasagna is called on startup
         if(input.h_rel > constants.equil_h_thresh){       // looks like we are already in flight ohp
           launch_h = 0;
           state.status = EQUIL;
         } else {
         launch_h = input.h_rel;
         }
+        is_first_call = false;
       } else if(input.h_rel - launch_h > constants.launch_h_thresh) state.status = ASCENT;
```

Comment:
Yeah even though it's wasn't from this PR im glad you looked at this, as the whole point of making this more readable is so that other people can catch edge cases and stuff. Anyways, you right about the behavior, however the launch_h_thresh defaults to 300m, so it shouldn't be an issue. launch_h_thresh should always be an small enough value such that it isn't significant. I can make a note in the header file where the default is set about why it needs to be relatively small.

Reply:
never apologize for asking questions without reading the code, you raise a good point. If the MCU reboots below the equilibration threshold (10km) while falling, then the controller will not kick in properly. I can fix that by having the threshold for detecting launch be an absolute value. I wont do that in this PR tho bc I would need to test that change separately, so I will do that in a later PR